### ISSUE TITLE
Change from replaced commit reappearing somewhere else

### ISSUE THREAD

**Issue by Code7R** [newren/git-filter-repo#368]
Hi,

I am using the current git-lfs-repo version from PIP. The intended task was

1. injecting a .gitattributes files into the begin of the history
2. making sure that this same file contents remains the same for the whole rebased tree

Part 1 is easy and works with the insert-begin script from examples, works fine for easy repositories.

Part 2 is not so easy. I have a commit burried deep in the history which only modified the .gitattributes files, and that change needs to be gone. So the natural way to get rid of the nasty commit was using replace-refs,
i.e. ` git replace unwanted-commit parent-of-the-unwanted-commit` 

And that does not work. There is no warning while processing but in the end the contents of .gitattributes is still replaced in the branch HEADs. But according to the history the replaced commit was actually wipped.

In fact, the commit which git-blame attriutes the last file change to is some totally unrelated commit which happened much later in the history. Actually that is a merge commit which should not happen any history whatsoever.

So I guess there is some bug in git-filter-repo which recreates zombie files from skipped commits triggered by some condition?

**Comment by newren** [comment]
No, this isn't a bug.  "Skipping commits" in filter-repo is almost always a bad idea; it just means the changes made from that commit tend to get squashed into a later commit.  I put it in there because filter-branch had such an ability, but it was a really bad idea for the same reasons in filter-branch; possibly even worse.  I believe the only place I documented commit skipping I put a large disclaimer with it -- see https://github.com/newren/git-filter-repo/blob/main/Documentation/converting-from-filter-branch.md#removing-commits-by-a-certain-author.

For more background, see https://github.com/newren/git-filter-repo/issues/62#issuecomment-597725502.  filter-repo, by virtue of using fast-export and fast-import, do not operate on a difference between commits, but on a "make these files have these contents" design.  Basically fast-export looks at which files are different from the first parent, and then re-issues the exact contents of those files.  For merge commits, that means it'll re-specify all files made on the branch being merged with their exact contents.  Thus the reason the changes from your "dropped commit" reappear later.

One way to do this would be to strip out the .gitattributes file from history and then run insert beginning; e.g.:
    git filter-repo --invert-paths --path .gitattributes
    insert_beginning --file .gitattributes
Alternatively, you could use a commit callback instead, and do this before or after running insert_beginning (warning: untested, I might have a typo or something but this should get you on the right path):
```
    git filter-repo --commit-callback '
        for change in commit.file_changes:
            if change.filename == b".gitattributes":
                change.sha = <KNOWN_GOOD_SHA>
    '
```
where <KNOWN_GOOD_SHA> would be something like `b"c0aab39e4817a1568f5a9890ab5733d1c5af8ce2"` (`git hash-object .gitattributes` before running git-filter-repo in a directory with the desired .gitattributes file can get you the necessary sha1sum).

Does that help?

**Comment by Code7R** [comment]
Hi, yes, that does help. I will probably go for the first option and eliminate .gitattributes first.
Also hack the insert_beginning script to extend the commit message and add a note to the initial commit about the reason of the modification.

Can be closed unless there is something more to discuss. Would be good to have some better warning in the manual, though.
